src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt

```diff
@@ -49,41 +112,21 @@ object RustResolveEngine {
             }
         }
 
-        override fun visitForExpr(o: RustForExpr) {
-            seek(o.scopedForDecl)
-        }
-
-        override fun visitScopedLetExpr(o: RustScopedLetExpr) {
-            visitResolveScope(o)
-        }
-
-        override fun visitLambdaExpr(o: RustLambdaExpr) {
-            visitResolveScope(o)
-        }
-
-        override fun visitMethod(o: RustMethod) {
-            visitResolveScope(o)
-        }
-
-        override fun visitFnItem(o: RustFnItem) {
-            visitResolveScope(o)
-        }
+        override fun visitForExpr(o: RustForExpr)               = seek(o.scopedForDecl)
+        override fun visitScopedLetExpr(o: RustScopedLetExpr)   = visitResolveScope(o)
+        override fun visitLambdaExpr(o: RustLambdaExpr)         = visitResolveScope(o)
+        override fun visitMethod(o: RustMethod)                 = visitResolveScope(o)
+        override fun visitFnItem(o: RustFnItem)                 = visitResolveScope(o)
+        override fun visitResolveScope(scope: RustResolveScope) = seek(scope.getDeclarations())
 
         override fun visitBlock(o: RustBlock) {
             o.getDeclarations()
                 .takeWhile { it.isBefore(name) }
+                .filterNot { PsiTreeUtil.isAncestor(it, name, true) }
```

Comment:
Please, note, that current version _does_ behave differently from the previous one: it fires `PsiTreeUtil.isAncestor` for every decl, while it definitely shouldn't.